src/Strings.h

```diff
@@ -0,0 +1,388 @@
+// ----------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License
+// ----------------------------------------------------------------------
+#pragma once
+
+namespace Microsoft {
+namespace Featurizer {
+namespace Details {
+
+template<typename IteratorT>
+inline std::vector<IteratorT> IteratorVectorGenerator(IteratorT const & begin,
+                                                     IteratorT const & end) {
+    std::vector<IteratorT> wordIterVector;
+    wordIterVector.reserve(static_cast<unsigned int>(std::distance(begin, end) + 1));
+
+    for (IteratorT strIter = begin; strIter != end; ++strIter)
+        wordIterVector.emplace_back(std::move(strIter));
+    wordIterVector.emplace_back(std::move(end));
+
+    return wordIterVector;
+}
+
+template <
+    typename IteratorT,
+    typename IsDelimiterT
+>
+inline void Parse(IteratorT const &begin,
+                  IteratorT const &end,
+                  IsDelimiterT const &isDelimiter,
+                  std::function<void (IteratorT&, IteratorT&)> const &callback) {
+
+    IteratorT left = begin;
+    IteratorT right = left;
+    while (left != end && *left == isDelimiter)
+        ++left;
+
+    right = left;
+    while (right != end) {
+        if (*right == isDelimiter) {
+            callback(left, right);
+            left = right;
+            while (left != end && *left == isDelimiter)
+                ++left;
+            right = left;
+        } else {
+            ++right;
+        }
+    }
+    if (left != right)
+        callback(left, right);
+}
+
+template <
+    typename IteratorT,
+    typename RegexT
+>
+inline void ParseRegex(IteratorT const &begin,
+                       IteratorT const &end,
+                       RegexT const &regexToken,
+                       std::function<void (IteratorT&, IteratorT&)> const &callback) {
+
+    std::sregex_iterator iter(begin, end, regexToken);
+    std::sregex_iterator iterEnd;
+
+    while (iter != iterEnd) {
+        std::string const iterStringConst = iter->str();
+        IteratorT beginRef = iterStringConst.begin();
+        IteratorT endRef = iterStringConst.end();
+        callback(beginRef, endRef);
+        ++iter;
+    }
+}
+
+template <typename IteratorT>
+inline std::string TrimAndReplace(IteratorT begin,
+                                  IteratorT end) {
+    //replace all punctuations with spaces
+    std::transform(begin, end, begin, [](char c) {
```

Comment:
>begin [](start = 31, length = 5)

This assumes two things: 1) The iterator is not const 2) the input iterator can be used as an output iterator. And this is all about strings. Is it possible to use standard string functions and standard algos?

Reply:
sure